### PR TITLE
Add agent for creating and partitioning devices

### PR DESCRIPTION
**Pull request by dluc**
# Description and Motivation 

Add a new agent to the solution, which will be responsible for creating the devices in IoT hub and for partitioning the list of devices, so that multiple VMs can each simulate a different set of devices.
The agent is not started yet because the new storage library is not ready.

The PR includes some config refactoring and cleanup, e.g. having all the storage settings in the same place, and moving the entire diagnostics url to config (instead of being partially hard coded) 

# Change type 

- [ ] Bug fix
- [x] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/251)

### CODE REVIEW

File: WebService/Startup.cs [hunk 1]
```diff
@@ -18,6 +19,10 @@ namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.WebService
 {
     public class Startup
     {
+        // Agent responsible for creating devices and partitions
+        private IPartitioningAgent partitioningAgent;
+
+        // Agent responsible for simulating IoT devicesw
```
[the remark this answers is not in the document]
Comment (dluc) [reply WebService/Startup.cs]: fixed

File: WebService/Startup.cs [hunk 2]
```diff
@@ -77,28 +82,34 @@ public IServiceProvider ConfigureServices(IServiceCollection services)
...
+            this.partitioningAgent = this.ApplicationContainer.Resolve<IPartitioningAgent>();
+
+            // TODO: uncomment when ready
+            //this.partitioningAgent.StartAsync();
+
             this.simulationAgent = this.ApplicationContainer.Resolve<ISimulationAgent>();
             this.simulationAgent.RunAsync();
         }
 
         private void StopAgent()
```
[the remark this answers is not in the document]
Comment (dluc) [reply WebService/Startup.cs]: done

File: PartitioningAgent/Agent.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        private bool running;
+
+        public Agent(
+            IClusterNodes cluster,
+            IThreadWrapper thread,
+            IClusteringConfig clusteringConfig,
+            ILogger logger)
+        {
+            this.clusterNodes = cluster;
```
Comment (peterfelts) [inline PartitioningAgent/Agent.cs]: Where does this get passed in from? If Autofac is doing it, I'm not clear on where the values are coming from.

File: Services/Clustering/ClusterNodes.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        public Task KeepAliveNodeAsync()
+        {
+            throw new NotImplementedException("Need the new storage code first");
+        }
+
+        // Generate a unique value used to identify the current instance
+        private static string GenerateSharedNodeId()
+        {
+            // Example: 12a34b5678901c23de4c5f6ab78c9012.2018-01-12T01:15:00
+            return Guid.NewGuid().ToString("N") + "." + DateTimeOffset.UtcNow.ToString("s");
```
Comment (peterfelts) [inline Services/Clustering/ClusterNodes.cs]: What's the reason for having the date added to the node ID? Is it just for readability?

File: PartitioningAgent/Agent.cs (hunk 3 above)
Comment (dluc) [reply PartitioningAgent/Agent.cs]: I don't understand the question, `cluster` is injected as usual, see line 26

File: Services/Clustering/ClusterNodes.cs (hunk 4 above)
Comment (dluc) [reply Services/Clustering/ClusterNodes.cs]: only for debugging purpose. When looking at logs containing node Id, each node will have a random id plus a datetime indicating when the node started

**Pull request merged**